File: extensions/sim-flow-vscode/src/llm/lmstudio.ts
// LM Studio backend. LM Studio's local server hosts an
// OpenAI-compatible endpoint at `<host>/v1/chat/completions`. The
// model id must match whatever is currently loaded in LM Studio,
// so we don't set a meaningful default - callers should supply
// `sim-flow.llm.model`.

import { OpenAiCompatibleBackend } from "./openai-compat";
import type { SecretStorage } from "./types";

export const LMSTUDIO_KEY_ID = "sim-flow.lmstudio.apiKey";
export const LMSTUDIO_DEFAULT_BASE_URL = "http://localhost:1234/v1";

export interface LMStudioBackendOptions {
  model?: string;
  modelFamilyId?: string;
  runtimeProfileId?: string;
  secrets?: SecretStorage;
  baseUrl?: string;
  apiUrl?: string;
  fetchImpl?: typeof fetch;
  /**
   * Override the backend's display name, used in error
   * diagnostics. The factory routes `vllm` and `openai-compat`
   * through this class (same wire format as LM Studio);
   * labeling them "lmstudio" in the error obscures the actual
   * selector the user picked.
   */
  name?: string;
}

export class LMStudioBackend extends OpenAiCompatibleBackend {
  constructor(options: LMStudioBackendOptions = {}) {
    super({
      name: options.name ?? "lmstudio",
      baseUrl: options.baseUrl ?? LMSTUDIO_DEFAULT_BASE_URL,
      defaultModel: "local-model",
      provider: "lmstudio",
      keyId: LMSTUDIO_KEY_ID,
      requireKey: false,
      model: options.model,
      modelFamilyId: options.modelFamilyId,
      runtimeProfileId: options.runtimeProfileId,
      secrets: options.secrets,
      apiUrl: options.apiUrl,
      fetchImpl: options.fetchImpl,
    });
  }
}

File: src/__internal/session/agent/lmstudio.rs
//! `LmStudioAgent` - LM Studio's local OpenAI-compatible endpoint.

use super::CliAgent;
use super::openai_compatible::{OpenAiCompatibleRequest, dispatch_chat};
use crate::Result;
use crate::session::protocol::LlmMessage;

pub const DEFAULT_BASE_URL: &str = "http://localhost:1234/v1";
/// LM Studio doesn't have a meaningful default - the user picks the
/// loaded model in the UI. We pass `local-model` as a placeholder
/// when nothing is configured; users normally set `--llm-model
/// <name>` to whatever is loaded.
pub const DEFAULT_MODEL: &str = "local-model";

pub struct LmStudioAgent {
    base_url: String,
    model: String,
}

impl LmStudioAgent {
    pub fn new(base_url: Option<String>, model: Option<String>) -> Self {
        Self {
            base_url: base_url.unwrap_or_else(|| DEFAULT_BASE_URL.into()),
            model: model.unwrap_or_else(|| DEFAULT_MODEL.into()),
        }
    }
}

impl CliAgent for LmStudioAgent {
    fn name(&self) -> &str {
        "lmstudio"
    }

    fn dispatch(&self, messages: &[LlmMessage]) -> Result<String> {
        dispatch_chat(OpenAiCompatibleRequest::new(
            &self.base_url,
            &self.model,
            messages,
        ))
    }
}
